Add a small example for use in a tutorial

shonfeder opened the pull request:
Contributes to #37

This PR introduces a small example showing use of the `apalache-chai` library. The material added here will be used in followup tutorial documentation, but it also serves (I hope) as a useful example on its own.

## Reviewing

Don't be too scared by the PR line count, as that includes a lockfile update.

This PR is best reviewed by commit, and only three commits are substantial:

- [2e0685b](https://github.com/informalsystems/apalache-chai/pull/38/commits/2e0685bf2c3ecd49723dd23ffb52954e80c7b9ff) introduces the TLA spec used for counter-example generation. (cc/ @Kukovec  and @bugarela and @thpani). Would love to learn ways to improve this, as I think it could be a very helpful basis for intro material more generally: it specifies a simple stateful system which every engineer knows, and could be expanded to include concurrent operations by different users etc.
- [f145c22](https://github.com/informalsystems/apalache-chai/pull/38/commits/f145c22de422c7f16f1bd98bd7a038f9d803b4d8) introduces the demo python app (cc/ @thpani and @hvanz). Aside from general code quality, the functions for manipulating the TLA ast and generating the graph can probably glossed over. But the RPC calls in the `main` function should be a good opportunity for the @informalsystems/mbt team to give input on changes they'd like in the library API.
- [9a5ea14](https://github.com/informalsystems/apalache-chai/pull/38/commits/9a5ea14996fa04b6427454e0903091e6284656b0) introduces a README to give a high level account of the example, and show what a session with it looks like. Feedback here on how to improve it's use as a brief pedagogical overview is welcome (but note that this is not mean to be an indepth tutorial at this point).

Thanks in advance for the review!

Commit:
Add demo FileSystem spec

Commit:
Add demo application using chai

Commit:
Add deps needed for example app

Commit:
Add README to example dir

Commit:
Fix import sorting

bugarela commented on example/app.py:
```diff
@@ -0,0 +1,242 @@
+
+
+def build_state_graph(counter_examples):
+    """Build a graph of all he states in the `counter_examples`"""
```
```suggestion
    """Build a graph of all the states in the `counter_examples`"""
```

bugarela commented on example/app.py:
```diff
@@ -0,0 +1,242 @@
+        set_model_params(args, model)
+        print("Model parameters updated from CLI params")
+
+        # Run the typechecker to obtain counterexamples
```
```suggestion
        # Run the model checker to obtain counterexamples
```

bugarela commented on example/app.py:
```diff
@@ -0,0 +1,242 @@
+                },
+            },
+        )
+        # NOTE: Production imlementations should include proper error handling
```
```suggestion
        # NOTE: Production implementations should include proper error handling
```

bugarela commented on example/app.py:
```diff
@@ -0,0 +1,242 @@
+
+        # Load the TLA into a JSON representation of the model
+        model = await client.typecheck(source)
+        # NOTE: Production imlementations should include proper error handling
```
```suggestion
        # NOTE: Production implementations should include proper error handling
```

bugarela commented on example/FileSystem.tla:
```diff
@@ -0,0 +1,138 @@
+\* A $cmd is a triple holding:
+\*
+\* - the command name
+\* - the path(s) it operators on
```
```suggestion
\* - the path(s) it operates on
```

bugarela reviewed (approved):
Interesting example! If we want to spend more time exploring visualization in the future, we could try to show the file tree structure in the node labels in the graph :) Also, if we parameterize the TLA+ Actions by filename, we could try to show those in the graph edges (i.e. `MkDir('foo')`).

Anyways, these are just ideas for the future. It is already a great example. Although users might miss a simpler straight forward example aside from this one, with: connect, typecheck, check, save itf files.

Commit:
Update example/app.py

Co-authored-by: Gabriela Moreira <gabrielamoreiramafra@gmail.com>

Commit:
Update example/app.py

Co-authored-by: Gabriela Moreira <gabrielamoreiramafra@gmail.com>

Commit:
Update example/app.py

Co-authored-by: Gabriela Moreira <gabrielamoreiramafra@gmail.com>

Commit:
Update example/app.py

Co-authored-by: Gabriela Moreira <gabrielamoreiramafra@gmail.com>

Commit:
Update example/FileSystem.tla

Co-authored-by: Gabriela Moreira <gabrielamoreiramafra@gmail.com>

shonfeder commented:
You're ideas for the visualization are great, and I'd wanted to do all of them actually (the commands do record the target path) but I found that fiddling with the graph was taking a lot of time, I opted for the more modest result you see here.

Very good point about the need for a more concise and minimal example. I think what we have in the API docs here should address that need?

https://github.com/informalsystems/apalache-chai/blob/98eef4f8bb6dae7110f58a9ec1d500841513a914/chai/cmd_executor.py#L144-L173

Thanks for the fixes and the review!

thpani commented on example/README.md:
```diff
@@ -0,0 +1,68 @@
+
+The demo is defined in [./app.py](./app.py). It performs the following actions:
+
+- Connects to a running apalche server
```
```suggestion
- Connects to a running Apalache server
```

thpani commented on example/README.md:
```diff
@@ -0,0 +1,68 @@
+The demo is defined in [./app.py](./app.py). It performs the following actions:
+
+- Connects to a running apalche server
+- Makes an RPC call to the server to parse a TLA file into a JSON representation
+  of the model
```
```suggestion
- Makes an RPC call to the server to parse a TLA+ file into a JSON representation
  of the model
```

thpani commented on example/README.md:
```diff
@@ -0,0 +1,68 @@
+- Listing the children of a directory  (like `ls`)
+
+The spec is used to generate counter-examples leading up to certain states. The
+python app takes CLI parameters that determine the properites of the states, and
```
```suggestion
python app takes CLI parameters that determine the properties of the states, and
```

thpani commented on example/FileSystem.tla:
```diff
@@ -0,0 +1,138 @@
+    \* The set of all files in the system.
+    \* @type: Set($file);
+    files,
+    \* The commend issued.
```
```suggestion
    \* The command issued.
```

thpani commented on example/FileSystem.tla:
```diff
@@ -0,0 +1,138 @@
+\*
+\* @typeAlias: path = Seq(Str);
+\* @typeAlias: file = File($path) | Dir($path);
+\* @type: $path => $file;
+File(path) == Variant("File", path)
+\* @type: $path => $file;
+Dir(path) == Variant("Dir", path)
```
I like this pattern of using constructory operators for variants!

thpani commented on example/FileSystem.tla:
```diff
@@ -0,0 +1,138 @@
+\* NOTE: The `Children` relation is reflexive: each file is its own child.
+\*
+\* @type: $file => Set($file);
+Children(p) ==
+  { child \in files: FilePath(p) = SubSeq(FilePath(child), 1, Len(FilePath(p))) }
```
I think you want to take the `min` of `Len(FilePath(p))` and `Len(FilePath(child))` as 3rd argument to `SubSeq`:

```
If [...] n > Len(seq), then the result is undefined.
```
https://apalache.informal.systems/docs/lang/sequences.html?highlight=SubSeq#subsequence

Kukovec replied:
Since you're invoking `FilePath(_)` multiple times, you can wrap it in LETs. Also, you can avoid constructing a new sequence by doing a domain `\A` instead of SubSeq.
```tla
LET PathOfP == FilePath(p) IN
{ child \in files: 
  LET PathOfC == FilePath(child) IN
  /\  Len(PathOfC) >= Len(PathOfP)
  /\ \A i \in DOMAIN PathOfP: PathOfP[i] = PathOfC[i]
}
```

shonfeder replied:
> you can avoid constructing a new sequence by doing a domain \A instead of SubSeq

Are there expected performance improvements by doing this? On the face of it, it looks like just a more low-level approach to checking a subseq, like doing a manual for loop instead of taking a slice.

shonfeder replied:
Fixed in 749fecde29951e32923710c5eb0f83115cbeba49

Thanks for the offline explanation of the benefits here!

thpani commented on example/app.py:
```diff
@@ -0,0 +1,242 @@
+
+This demo performs the following actions:
+
+- Connects to a running apalche server
+- Makes an RPC call to the server to parse a TLA file into a JSON representation
+  of the model
```
```suggestion
- Connects to a running Apalache server
- Makes an RPC call to the server to parse a TLA+ file into a JSON representation
  of the model
```

thpani reviewed (approved):
Very nice, clearly written example! 🎉

I think your use of `SubSeq` can be undefined.
Otherwise just a few of typos.

Kukovec commented on example/FileSystem.tla:
```diff
@@ -0,0 +1,138 @@
+
+VARIABLES
+    \* The set of all files in the system.
+    \* @type: Set($file);
```
I'm surprised this works, the alias is defined after it's used.

Kukovec commented on example/FileSystem.tla:
```diff
@@ -0,0 +1,138 @@
+\* ACTIONS *\
+\***********\
+
+\* @type: Bool;
+Touch ==
+  \E d \in files:
+    /\ IsDir(d)
+    /\ \E n \in Names:
+      \* @type: $file;
+      LET newFile == File(Append(FilePath(d), n)) IN
+      /\ files' = { newFile } \union files
+      /\ cmd' = TouchCmd(newFile)
```
Not sure if you want to consider this or not, but shouldn't this also include a check for whether a file with that name, in that path already exists? Maybe have an error variable to track whether the last operation succeeds?
I'd make `Touch` take `d` and `name` as parameters too, feels more intuitive:
```tla
\* @type: ($file, Str) => Bool;
Touch(d, name) ==
    /\ IsDir(d)
    /\ LET 
       \* @type: $file;
       newFile == File(Append(FilePath(d), n)) 
       IN
       /\ error' = newFile \in files
       /\ files' = { newFile } \union files \* If error' = true this is a no-op
       /\ cmd' = TouchCmd(newFile)
```
`MkDir` would be similar, then next would be
```tla
\E d \in files:
  \/ \E n \in Names:
    \/ Touch(d, n)
    \/ MkDir(d,n)
  \/ Ls(d)
```

shonfeder replied:
Fixed, thanks! b5c28f8ffefc665f92d0a413becf3a04bd905058

Kukovec reviewed (commented):
Disclaimer, I just read the TLA file.

Commit:
Update example/README.md

Co-authored-by: Thomas Pani <thomas.pani@gmail.com>

Commit:
Update example/README.md

Co-authored-by: Thomas Pani <thomas.pani@gmail.com>

Commit:
Update example/README.md

Co-authored-by: Thomas Pani <thomas.pani@gmail.com>

Commit:
Update example/FileSystem.tla

Co-authored-by: Thomas Pani <thomas.pani@gmail.com>

Commit:
Add `IsPrefix` operator

Commit:
Make actions parameterized

Following guidance from @kukovec on best practices in structuring action
data dependencies and explanation of the advantages of prenex form.

Commit:
Fix typo

shonfeder commented:
Thanks for the reviews and feeback, all! <3

Commit:
Link README to example app
